File: String Matching/KMP Algorithm/kmp.java
 public static int[] kmpAlgo(String s) {
        int n = s.length();
        int[] lps = new int[n];

        for (int i = 1; i < n; i++) {
            int prevIdx = lps[i - 1];

            while (prevIdx > 0 && s.charAt(i) != s.charAt(prevIdx)) {
                prevIdx = lps[prevIdx - 1];
            }

            lps[i] = prevIdx + (s.charAt(i) == s.charAt(prevIdx) ? 1 : 0);
        }

        return lps;
    }

File: String Matching/KMP Algorithm/kmp.js
function kmpAlgo(s) {
  const n = s.length;
  const lps = new Array(n).fill(0);

  for (let i = 1; i < n; i++) {
      let prevIdx = lps[i - 1];

      while (prevIdx > 0 && s[i] !== s[prevIdx]) {
          prevIdx = lps[prevIdx - 1];
      }

      lps[i] = prevIdx + (s[i] === s[prevIdx] ? 1 : 0);
  }

  return lps;
}